corgidrp/calibrate_kgain.py

```diff
@@ -0,0 +1,918 @@
+    
+    # define bounds for linearity fit
+    # lower bound indices should include as many data points as possible
+    # upper bound indices should avoid nonlin region
+    lower_bound = np.min(np.where(compiled_binned_averages > min_val)[0])
+    upper_bound = np.max(np.where(compiled_binned_averages < max_val)[0])
+    
+    # Logarithmic transformation of specific array segments
+    logged_averages = np.log10(compiled_binned_averages[lower_bound:upper_bound])
+    logged_deviations = np.log10(compiled_binned_shot_deviations[lower_bound:upper_bound])
+    
+    # Find indices of NaN's
+    nan_dev = np.isnan(logged_deviations)
+    nan_avg = np.isnan(logged_averages)
+    
+    # Filter out NaN's from the data
+    logged_deviations = logged_deviations[~nan_dev & ~nan_avg]
+    logged_averages = logged_averages[~nan_dev & ~nan_avg]
+    
+    # Calculate the error factor
+    logged_deviations_error_factor = compiled_binned_deviations_error / compiled_binned_shot_deviations
+    logged_deviations_error_factor = logged_deviations_error_factor[lower_bound:upper_bound]
+    logged_deviations_error_factor = logged_deviations_error_factor[~nan_dev & ~nan_avg]
+    
+    # Calculate the logged deviations error
+    logged_deviations_error = logged_deviations_error_factor * logged_deviations
+    
+    # Create a mask for non-NaN values
+    non_nan_mask = ~np.isnan(logged_deviations_error)
+
+    # Use the mask to filter out NaN values
+    y_err = logged_deviations_error[non_nan_mask]
+    x_vals = logged_averages[non_nan_mask]
+    y_vals = logged_deviations[non_nan_mask]
+    
+    # make array of kgain values from variance and mean values
+    # equation: k_gain = mean / signal variance
+    kgain_arr = 10**(x_vals)/(10**y_vals)**2
+    
+    if make_plot:
+        fname = 'kgain_histogram'
+        plt.figure()
+        # 'auto' lets matplotlib decide the number of bins
+        plt.hist(kgain_arr, bins='auto', log=True)
+        plt.title('Histogram of kgain values')
+        plt.savefig(f'{plot_outdir}/{fname}')
+        if verbose:
+            print(f'Figure {fname} stored in {plot_outdir}')
+        if show_plot:
+            plt.show()
+        plt.close()
+    
+    if make_plot:
+        fname = 'kgain_vs_mean'
+        plt.figure()
+        plt.plot(10**x_vals, kgain_arr, marker='o', linestyle='-', color='b', label='kgain')
+        # Set y-axis range
+        plt.ylim(6, 11)
+        # Add a horizontal line at y = 10
+        plt.axhline(y=8.7, color='r', linestyle='--', label='y = 8.7')
+        # Add a grid
+        plt.grid(True)
+        plt.title('kgain versus mean')
+        plt.savefig(f'{plot_outdir}/{fname}')
+        if verbose:
+            print(f'Figure {fname} stored in {plot_outdir}')
+        if show_plot:
+            plt.show()
+        plt.close()
+    
+    kgain_clipped, _ = sigma_clip(kgain_arr)
+    mode_kgain,_ = calculate_mode(kgain_clipped)
```

Comment:
I think the `err` for the KGain class should be `np.std(kgain_clipped)` and not what it currently is in line 812.

Reply:
Also, I can't remember if we discussed this or not already, but I think the read noise (and its uncertainty would be nice, too) should be saved as header info in the `KGain` instance that is the output of this function.  We could probably discuss this at the meeting tomorrow.

Reply:
Thanks @kjl0025! First point done. UT run and OK. As I shared with you `err` was before your suggestion `mean_rn_gauss_DN` = np.mean(rn_gauss), where `rn_gauss = [Single_peakfit(bin_locs,counts_diff[x])/np.sqrt(2) for x in range(len(frames_diff))]` and `Single_peakfit' is just std(): "Fit Gauss function to x, y data. Returns mean and sigma. Only sigma
    is used in main code call"

Second one: https://github.com/roman-corgi/corgidrp/issues/135

Reply:
Sorry, I don't understand (concerning the first point).  I see that `err` in line 812 is the read noise estimate you just mentioned, but that is not the standard deviation of the array `kgain_clipped`.

Reply:
I see that was just changed in the latest commit.  Great, I'll resolve this conversation.